src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -309,49 +309,55 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
 
     /// Get the metadata for a target in a specific profile
     pub fn target_metadata(&self, unit: &Unit) -> Option<Metadata> {
-        let metadata = unit.target.metadata();
+        // No metadata for dylibs because of a couple issues
+        // - OSX encodes the dylib name in the executable
+        // - Windows rustc multiple files of which we can't easily link all of them
+        if !unit.profile.test && unit.target.is_dylib() {
```

Comment:
> How to handle the upstream dylib case

I think the conditions we're going to want here are:

```rust
if !unit.profile.test && 
   unit.target.is_dylib() && 
   unit.pkg.package_id().source_id().is_path() &&
   magical_env_var_not_set() {
    return None
}
```

That is, we want metadata on all tests, we want metadata on everything that's not a dylib, we want metdata on all upstream deps (non-path deps), and finally we have this magical env var thing.

Note that "upstream" == "not a path dep" because the current package changes when you move around a workspace. A path dep is basically a heuristic, but it works fine for us.

So for the magical env var, the problem is that if we *didn't* have that clause in the condition then when we `cargo build` the standard library we get a bunch of files that are literally `libstd.so`, `librustc.so`, etc. Now we don't want this because it means when you install Rust on your system you'd get `/usr/lib/libstd.so`. That's a little too "official"  for us to take! To solve this problem we've always added some form of hash to the libraries, so you'd instead install `/usr/lib/libstd-xxxxxxxx.so`. 

So basically when we use Cargo to build the standard library we want to unconditionally insert all these hashes, so the env var just says "please insert hashes everywhere".

Now if you're curious why we don't *always* install the hash, that's also interesting. The main idea is that if I have a project called "foo" and I built a dylib, we want to generate the predictable filename `target/debug/libfoo.so`. (no hash). Now for this to actually work correctly, however, we can't take the same trick we're doing with staticlibs and executables where we just hard link the file to the correct name. Platforms like OSX use the generate file name (`target/debug/deps/libfoo-xxxxxx.so`) at runtime for various reasons, so we need to generate the file as literally `target/debug/deps/libfoo.so` for now. We can fix this with linker options but rustc doesn't expose that yet (and is a problem for another day!)

Does that all make sense?